### PR TITLE
Implement S4060 - Avoid unsealed attributes - for VB.NET

### PR DESCRIPTION
**Pull request by Corniel**
See #501

# Description

The .NET framework class library provides methods for retrieving custom attributes. Sealing the attribute eliminates the search through the inheritance hierarchy, and can improve performance.

This rule raises an issue when a public type inherits from System.Attribute, is not abstract, and is not sealed.
## Noncompliant Code Example

``` VB
Namespace MyLibrary 

    <AttributeUsage(AttributeTargets.Class)>
    Public Class MyAttribute ' Noncompliant
        Inherits Attribute 
    
        Public Sub New(name as String) 
            Value = val
        End Sub
        
        Public Readonly Property Value as String
    End Class
End Namespace
```

## Compliant Solution

``` VB
Namespace MyLibrary 

    <AttributeUsage(AttributeTargets.Class)>
    Public MustInherit Class MyAttribute ' Noncompliant
        Inherits Attribute 
    
        Public Sub New(val as String) 
            Value = val 
        End Sub
        
        Public Readonly Property Value as String
    End Class
End Namespace
```

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel can you please rebase all your PRs on master? To remove the `RspecStrings.resx` files. It will help me to add the required resources around rspec.

**Comment by pavel-mikula-sonarsource** [comment]
I've pushed a rspec scaffolding commit here. Please pull it before rebase 😃

**Comment by Corniel** [comment]
Rebase done.

File: analyzers/rspec/vbnet/S4060_vb.net.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    Public Sub New(Name As String)
+        Me.Name = Name
+    End Sub
+
+End Class
+</pre>
+<h2>Compliant Solution</h2>
+<pre>
+Public NotInheritable Class MyAttribute    ' Noncompliant
```
Comment (Corniel) [inline analyzers/rspec/vbnet/S4060_vb.net.html:19]: ' Noncompliant for the compliant part.

File: analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                if (Language.Syntax.NodeIdentifier(c.Node) is { } identifier
+                    && !identifier.IsMissing
+                    && c.SemanticModel.GetDeclaredSymbol(c.Node) is INamedTypeSymbol { IsAbstract: false, IsSealed: false } symbol
+                    && symbol.DerivesFrom(KnownType.System_Attribute)
+                    && symbol.IsPubliclyAccessible())
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, identifier.GetLocation()));
+                }
+            },
+            Language.SyntaxKind.ClassAndRecordDeclaration);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs:49]: Record cannot inherit attribute and we should not register for it. You'll need to add `ClassDeclaration` to the facade
```suggestion
            Language.SyntaxKind.ClassDeclaration);
```
```suggestion
            Language.SyntaxKind.ClassDeclaration);
```

File: analyzers/rspec/vbnet/S4060_vb.net.html (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/rspec/vbnet/S4060_vb.net.html]: Fixed in rspec repo too

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Looks good, I left some minor polishing comments

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AvoidUnsealedAttributes.vb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿Public Class PublicAttribute ' Noncompliant {{Seal this attribute or make it abstract.}}
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AvoidUnsealedAttributes.vb:1]: There's a problem with the way how our test framework works. We can't assert locations `^^^^` on the first line.
Moving the problem to 2nd line will fix the issue.
```suggestion
' Test framework can't assert location on 1st line
﻿Public Class PublicAttribute ' Noncompliant {{Seal this attribute or make it abstract.}}
```

File: analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    protected AvoidUnsealedAttributesBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterSyntaxNodeActionInNonGenerated(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (Language.Syntax.NodeIdentifier(c.Node) is { IsMissing: false } identifier
+                    && c.SemanticModel.GetDeclaredSymbol(c.Node) is INamedTypeSymbol { IsAbstract: false, IsSealed: false } symbol
+                    && symbol.DerivesFrom(KnownType.System_Attribute)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs:42]: The coverage isn't great. Can you add this to cover the `DerivesFrom` scenario?
```
public class NotAnAttribute { }
```

File: analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    protected override string MessageFormat => "Seal this attribute or make it abstract.";
+
+    protected AvoidUnsealedAttributesBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterSyntaxNodeActionInNonGenerated(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (Language.Syntax.NodeIdentifier(c.Node) is { IsMissing: false } identifier
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs:40]: Also this to cover the `IsMissing`
```
public class { }    // Error CS1001  Identifier expected
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**